### PR TITLE
Better argspecs for Axes.stem and axes3d.plot_surface

### PR DESCRIPTION
**Pull request by Zac-HD**
Another entry in my long quest for explicit argspecs for Matplotlib.  Now with bonus fix for an unrelated test, because I happened to see it.

- [x] Code is PEP 8 compliant
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: doc/api/next_api_changes/2018-05-06-ZHD.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Consistent handling of *args in Axes.stem
+-----------------------------------------
+
+:meth:`matplotlib.axex.Axes.stem` now raises TypeError when passed
+unhandled positional arguments.  If two or more arguments are passed
+(ie X, Y, [linefmt], ...) and Y cannot be cast to an array, an error\
```
Comment (WeatherGod) [inline doc/api/next_api_changes/2018-05-06-ZHD.rst]: extraneous backslash

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -2350,18 +2350,20 @@ def stem(self, *args, linefmt=None, markerfmt=None, basefmt=None,
...
         y = np.asarray(args[0])
         args = args[1:]
 
         # Try a second one
         try:
-            x, y = y, np.asarray(args[0], dtype=float)
-        except (IndexError, ValueError):
-            # The second array doesn't make sense, or it doesn't exist
+            x, *args = args
+        except IndexError:
```
Comment (WeatherGod) [inline lib/matplotlib/axes/_axes.py]: There isn't any indexing anymore, or is this what python raises when it can't unpack?

File: lib/matplotlib/tests/test_mlab.py [hunk 3]
```diff
@@ -1131,7 +1131,7 @@ def test_detrend_str_linear_2d_slope_off_axis0(self):
         res = mlab.detrend(input, key='linear', axis=1)
         assert_allclose(res, targ, atol=self.atol)
 
-    def test_detrend_detrend_linear_1d_slope_off_axis1(self):
+    def test_detrend_detrend_linear_1d_slope_off_axis1_notranspose(self):
```
Comment (WeatherGod) [inline lib/matplotlib/tests/test_mlab.py]: Was there a particular need to change these test names?

Comment (Zac-HD) [reply lib/matplotlib/tests/test_mlab.py]: Yes - there's another test with this name, so before renaming this test could never be run!

**Comment by Zac-HD** [comment]
Thanks for the tip (and early review) @WeatherGod; I've just taken out the `plot_surface` part entirely for now and can revisit after the other PR is done.

**Comment by Zac-HD** [comment]
Hi @tacaswell @anntzer - can I get a quick review for this 15-line PR?

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -2350,18 +2350,20 @@ def stem(self, *args, linefmt=None, markerfmt=None, basefmt=None,
...
         # Assume there's at least one data array
         y = np.asarray(args[0])
         args = args[1:]
 
         # Try a second one
         try:
-            x, y = y, np.asarray(args[0], dtype=float)
-        except (IndexError, ValueError):
-            # The second array doesn't make sense, or it doesn't exist
+            x, *args = args
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: That seems to be a slightly convoluted way to do `if len(args) > 1` (yes, I know, the original version is just as bad).

Moreover it is far from obvious to me that we want to drop support for passing just `y` (in which case `x` is implicitly defined as `range(len(y))`).  Yes, it's a questionable API, but it's also present in `plot()` and I don't think it makes sense to remove it in a piecemeal fashion; instead this should be done in a consistent manner.

Of course, there's the additional complication of categorical data (although I think categorical y's in a stem plot don't make much sense either).

TLDR: I don't think the API change is as benign as the PR suggests, and warrants more discussion.

Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: Agreed, a check `if len(args) > 1` is preferable to this `try` logic.

Support for `y`-only is not dropped. See the except clause.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -2350,17 +2350,19 @@ def stem(self, *args, linefmt=None, markerfmt=None, basefmt=None,
             which inspired this method.
 
         """
+        if len(args) > 5:
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: You may as well do
```
if not 1 <= len(args) <= 5:
    raise TypeError("stem expected between 1 and 5 positional arguments, got {}".format(len(args))
```
which lets you get rid of the comment immediately below.

Comment (Zac-HD) [reply lib/matplotlib/axes/_axes.py]: Done.

**Comment by Zac-HD** [comment]
> modulo CI

🎉
